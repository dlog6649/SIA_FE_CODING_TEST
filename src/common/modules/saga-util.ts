import { put } from "redux-saga/effects"

export const initAsyncState: AsyncState<any> = {
  isLoading: false,
  data: null,
  error: null,
}

export type AsyncState<D> = {
  isLoading: boolean
  data: D | null
  error: Error | null
}

export enum AsyncSuffix {
  Loading = "Loading",
  Success = "Success",
  Failure = "Failure",
}

export function taker(type: string, fn: (action: any) => Promise<any>, sagaOption: any) {
  return sagaOption(type, function* (action: any) {
    yield put({ type: type + AsyncSuffix.Loading })
    try {
      const response = yield fn(action)
      yield put({
        type: type + AsyncSuffix.Success,
        payload: response.data ? response.data : response,
      })
    } catch (err) {
      yield put({
        type: type + AsyncSuffix.Failure,
        payload: err,
      })
    }
  })
}

export const handleAsyncAction = (state: any, action: any) => {
  // ex) action.type: someReducer/getSomeDataLoading
  const [reducerName, type] = action.type.split("/")
  if (type.includes(AsyncSuffix.Loading)) {
    const api = getApiString(type, AsyncSuffix.Loading)
    state.api[api] = {
      isLoading: true,
      data: getData(state, api),
      error: null,
    }
  } else if (type.includes(AsyncSuffix.Success)) {
    const api = getApiString(type, AsyncSuffix.Success)
    console.log(type, api, action)
    state.api[api] = {
      isLoading: false,
      data: action.payload,
      error: null,
    }
  } else if (type.includes(AsyncSuffix.Failure)) {
    const api = getApiString(type, AsyncSuffix.Failure)
    state.api[api] = {
      isLoading: false,
      data: getData(state, api),
      error: action.payload,
    }
  }
}
const getApiString = (type: string, suffix: AsyncSuffix) => type.substring(0, type.length - suffix.length)
const getData = (state: any, api: string) => state.api[api]?.data || null
