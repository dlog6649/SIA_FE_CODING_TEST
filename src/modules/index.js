import { combineReducers } from 'redux';
import annotator from './annotator';

const rootReducer = combineReducers({
    annotator
});

export default rootReducer;