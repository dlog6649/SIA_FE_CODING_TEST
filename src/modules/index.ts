import { combineReducers } from "redux";
import annotator from "./annotator";

const rootReducer = combineReducers({
  annotator,
});

export default rootReducer;

export type RootState = ReturnType<typeof rootReducer>;
