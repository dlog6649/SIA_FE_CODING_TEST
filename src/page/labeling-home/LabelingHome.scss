.labeling-home {
  padding: 50px 100px;
  >.title {
    padding: 35px 15px;
    font-size: 2.5rem;
    font-weight: bold;
    color: #0f1350;
  }
  >.content {

  }
}
