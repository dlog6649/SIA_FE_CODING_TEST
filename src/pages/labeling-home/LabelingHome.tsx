import Card from "@src/components/card/Card"
import { useImagesQuery } from "@src/domains/image/queries"
import type { Image } from "@src/domains/image/types"
import Paths from "@src/shared/Paths"
import React from "react"
import { Link } from "react-router-dom"

import styles from "./LabelingHome.module.scss"

export default function LabelingHome() {
  const { isFetching, isError, data: images } = useImagesQuery()

  return (
    <main className={styles.labelingHome}>
      <h1 className={styles.title}>{"Labeling Home"}</h1>
      <div className={styles.cardItemBox}>
        {isFetching ? (
          <h2>{"Loading..."}</h2>
        ) : isError ? (
          <h2>{"Error occurred when fetching images"}</h2>
        ) : !images?.length ? (
          <h2>{"No Data"}</h2>
        ) : (
          images.map((img: Image) => (
            <Link key={img.id} className={styles.imgCard} to={Paths.buildLabelingBoardNav(img.id)}>
              <Card thumbnailUrl={img.thumbnailUrl} text={img.title} />
            </Link>
          ))
        )}
      </div>
    </main>
  )
}
