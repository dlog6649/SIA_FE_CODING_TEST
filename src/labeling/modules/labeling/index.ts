import { createAction, createReducer } from "@reduxjs/toolkit"
import { AsyncSuffix, handleAsyncAction, initAsyncState } from "../../../common/modules/util"
import { LabelingState } from "./types"

export const GET_IMAGES = "labeling/getImages"
const GET_IMAGES_LOADING = `${GET_IMAGES}/${AsyncSuffix.Loading}`
const GET_IMAGES_SUCCESS = `${GET_IMAGES}/${AsyncSuffix.Success}`
const GET_IMAGES_FAILURE = `${GET_IMAGES}/${AsyncSuffix.Failure}`
export const getImages = createAction<void>(GET_IMAGES)

export const GET_IMAGE = "labeling/getImage"
const GET_IMAGE_LOADING = `${GET_IMAGE}/${AsyncSuffix.Loading}`
const GET_IMAGE_SUCCESS = `${GET_IMAGE}/${AsyncSuffix.Success}`
const GET_IMAGE_FAILURE = `${GET_IMAGE}/${AsyncSuffix.Failure}`
export const getImage = createAction<string>(GET_IMAGE)

export const initState: LabelingState = {
  api: {
    getImages: initAsyncState,
    getImage: initAsyncState,
  },
}

export const labelingReducer = createReducer(initState, {
  [GET_IMAGES_LOADING]: (state, action) => handleAsyncAction(state, action),
  [GET_IMAGES_SUCCESS]: (state, action) => {
    action.payload = action.payload.slice(0, 12)
    handleAsyncAction(state, action)
  },
  [GET_IMAGES_FAILURE]: (state, action) => handleAsyncAction(state, action),
  [GET_IMAGE_LOADING]: (state, action) => handleAsyncAction(state, action),
  [GET_IMAGE_SUCCESS]: (state, action) => handleAsyncAction(state, action),
  [GET_IMAGE_FAILURE]: (state, action) => handleAsyncAction(state, action),
})

export * from "./types"
export * from "./saga"
