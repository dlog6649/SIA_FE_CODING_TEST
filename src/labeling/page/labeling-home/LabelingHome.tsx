import React, { useEffect } from "react"
import { Link } from "react-router-dom"
import { useDispatch } from "react-redux"
import styles from "./LabelingHome.module.scss"
import * as routes from "../../../routes"
import { getImages, Image } from "../../modules/labeling"
import Card from "../../../common/components/card/Card"
import { useRootState } from "../../../common/hooks/useRootState"

export default function LabelingHome() {
  const { loading, data, error } = useRootState((state) => state.labelingReducer.api.getImages)
  const dispatch = useDispatch()

  useEffect(() => {
    dispatch(getImages())
    // dispatch(getImages())
  }, [])

  return (
    <main className={styles.labelingHome}>
      <h1 className={styles.title}>{"Labeling Home"}</h1>
      <div className={styles.cardItemBox}>
        {loading ? (
          <h2>{"Loading..."}</h2>
        ) : error ? (
          <h2>{error.toString()}</h2>
        ) : !data?.length ? (
          <h2>{"No Data"}</h2>
        ) : (
          data.map((img: Image) => (
            <Link className={styles.imgCard} to={routes.buildLabelingViewPath(img.id)} key={img.id}>
              <Card thumbnailUrl={img.thumbnailUrl} text={img.title} />
            </Link>
          ))
        )}
      </div>
    </main>
  )
}
