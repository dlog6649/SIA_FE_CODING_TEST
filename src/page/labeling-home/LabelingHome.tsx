import React, { useEffect } from "react";
import CardItemBoxContainer from "../../container/CardItemBoxContainer";
import "./LabelingHome.scss";

export default function LabelingHome() {
  const title = "Labeling Home";
  useEffect(() => {
    console.log("Labeling useEffect");
  });
  return (
    <div className="labeling-home">
      <div className="title">{title}</div>
      <div className="content">
        <CardItemBoxContainer />
      </div>
    </div>
  );
}
