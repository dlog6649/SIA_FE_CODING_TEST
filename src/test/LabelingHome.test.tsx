import LabelingHome from "../labeling/page/labeling-home/LabelingHome"
import { Provider, useDispatch, useSelector } from "react-redux"
import * as reactRedux from "react-redux"
import React from "react"
import { render, fireEvent, waitFor, RenderResult } from "@testing-library/react"
import configureMockStore from "redux-mock-store"
import * as labelingModule from "../labeling/modules/labeling"

const mockStore = configureMockStore()
const store = mockStore({ labelingReducer: labelingModule.initState })

describe("Test LabelingHome", () => {
  const useSelectorMock = jest.spyOn(reactRedux, "useSelector")
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch")

  // const mockViewImage = jest.fn()
  // jest.setTimeout(30000)

  beforeEach(() => {
    // useSelectorMock.mockClear()
    // useDispatchMock.mockClear()
  })
  jest.mock("react-redux")

  // it("123", () => {
  //   useSelectorMock.mockReturnValue({ aa: "aa" })
  //   const dummySelector = jest.fn()
  //   const dummyDispatch = jest.fn()
  //   useDispatchMock.mockReturnValue(dummyDispatch)
  //
  //   expect(dummyDispatch).not.toHaveBeenCalled()
  // })

  it("ㅇㅇ", () => {
    // useSelectorMock.mockReturnValue({ aa: "aa" })
    // const dummySelector = jest.fn()
    // const dummyDispatch = jest.fn()
    // useDispatchMock.mockReturnValue(dummyDispatch)
    // expect(dummyDispatch).not.toHaveBeenCalled()
    const wrapper = render(
      <Provider store={store}>
        <LabelingHome />,
      </Provider>,
    )
    // expect(useDispatchMock).toHaveBeenCalledTimes(10)
    expect(useDispatch).toHaveBeenCalledWith()
  })

  // jsonplaceholder 비동기 대기시간으로 30초 부여
  // await waitFor(() => wrapper.getAllByTestId("testThumbnail"), { timeout: 30000 })
  //   .then(() => {
  //     expect(wrapper.baseElement).toMatchSnapshot()
  //
  //     const testThumbnails = wrapper.getAllByTestId("testThumbnail")
  //     expect(testThumbnails.length).toBe(8)
  //
  //     // 0번 썸네일 클릭
  //     fireEvent.click(testThumbnails[0])
  //
  //     const viewImageEventByClick = mockViewImage.mock.calls[0]
  //     const [url, title] = viewImageEventByClick
  //     expect(url).toEqual("https://via.placeholder.com/600/92c952")
  //     expect(title).toEqual("accusamus beatae ad facilis cum similique qui sunt")
  //   })
  //   .catch((err) => console.log(err))
})
