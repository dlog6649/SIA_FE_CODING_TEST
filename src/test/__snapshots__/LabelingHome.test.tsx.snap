// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelingHome 테스트 시작 썸네일을 클릭 시 해당 정보가 디스패치돼야 한다. 1`] = `
<body>
  <div>
    <main
      class="labelingHome"
    >
      <h1
        class="title"
      >
        Labeling Home
      </h1>
      <div
        class="cardItemBox"
      >
        <h2>
          No Data
        </h2>
      </div>
    </main>
    ,
  </div>
</body>
`;
